### PR TITLE
Fix: Fixed an issue where it didn't work to map certain keys to Actions

### PR DESCRIPTION
**Pull request by XTorLukas**
**Resolved / Related Issues**

To prevent extra work, all changes to the Files codebase must link to an approved issue marked as `Ready to build`. Please insert the issue number following the hashtag with the issue number that this Pull Request resolves.
- #15403

**Steps used to test these changes**

Stability is a top priority for Files and all changes are required to go through testing before being merged into the repo. Please include a list of steps that you used to test this PR.

1. Testing the correct mapping of all problem keys

### CODE REVIEW

File: src/Files.App/Data/Commands/HotKey/HotKey.cs [hunk 1]
```diff
@@ -332,9 +332,21 @@ private static string GetLocalizedKey(string key)
...
 		{
-			var buffer = new StringBuilder(256);
+			var buffer = new StringBuilder(2);
 			var state = new byte[256];
-			_ = Win32PInvoke.ToUnicode((uint)key, 0, state, buffer, 256, 0);
+
+			if (!Win32PInvoke.GetKeyboardState(state))
+			{
+				return buffer.ToString();
+			}
```
Comment (0x5bfa) [inline src/Files.App/Data/Commands/HotKey/HotKey.cs:341]:
```suggestion
			if (!Win32PInvoke.GetKeyboardState(state))
				return buffer.ToString();
```

File: src/Files.App/Data/Commands/HotKey/HotKey.cs [hunk 2]
```diff
@@ -332,9 +332,21 @@ private static string GetLocalizedKey(string key)
...
+
+			var virtualKey = (uint)key;
+			var scanCode = Win32PInvoke.MapVirtualKey(virtualKey, 0);
+
+			if (Win32PInvoke.ToUnicode(virtualKey, scanCode, state, buffer, buffer.Capacity, 0) > 0)
+			{
+				return buffer[^1].ToString();
+			}
```
Comment (0x5bfa) [inline src/Files.App/Data/Commands/HotKey/HotKey.cs:349]:
```suggestion
			if (Win32PInvoke.ToUnicode(virtualKey, scanCode, state, buffer, buffer.Capacity, 0) > 0)
				return buffer[^1].ToString();
```

**Review by 0x5bfa — APPROVED** [review]
You know our codebase well!!!
LGTM. code wise.
As we are not in an area that needs unicode support, could you please send a screenshot or video recording of the validation?

**Comment by XTorLukas** [comment]
@yaira2 I found the cause of the NumKeys mapping not working, Not localized Pad0-9
image
image

**Comment by XTorLukas** [comment]
### Oh, Decimal (NumPad) key not defined too
image

**Comment by XTorLukas** [comment]
### Test own layout (cs-CZ_QWERTZ)
Qwertz_cz svg (2)

* **All** Green keys are correct
* **Yellow** keys unused
* **Red keys** (Caps Lock / Scroll Lock / Num Lock) would be good to disable in the selection, because they are not possible, but they are marked as Empty string and so can be overlooked. After saving the application fails.

**Comment by XTorLukas** [comment]
> ### Test own layout (cs-CZ_QWERTZ)
> Qwertz_cz svg (2)
> 
> * **All** Green keys are correct
> * **Yellow** keys unused
> * **Red keys** (Caps Lock / Scroll Lock / Num Lock) would be good to disable in the selection, because they are not possible, but they are marked as Empty string and so can be overlooked. After saving the application fails.

### Now ignored non-valid keys aka **Red keys** and more optimalization for speed

**Comment by XTorLukas** [comment]
### Let's get to testing
- [ ] Testing functionality across different locations
  - [ ] cs-CZ_QWERTZ Layout
  - [ ] en-US_QWERTY Layout
  - [ ]  ... more


- [ ] Verification and search for possible bugs
- [ ] Finalization and verification

**Comment by XTorLukas** [comment]
I added information about invalid key binding to give the user the option to use a different key binding.
image

**Comment by XTorLukas** [comment]
### Let's get to testing
- [ ] Testing functionality across different locations
  - [x] `cs-CZ`
  - [x] `sk-SK`
  - [ ] `en-US`
  - [ ]  ... more


- [x] Verification and search for possible bugs
- [ ] Finalization

File: src/Files.App/Views/Settings/ActionsPage.xaml [hunk 3]
```diff
@@ -410,5 +411,13 @@
 			PreferredPlacement="Auto"
 			Subtitle="{helpers:ResourceString Name=KeybindingAlreadyUsedNotification}" />
 
+		<!--  Invalid Key TeachingTip  -->
+		<TeachingTip
```
Comment (yaira2) [inline src/Files.App/Views/Settings/ActionsPage.xaml:415]: Can you send a screenshot? Also, is there a way I can display this?

Comment (XTorLukas) [reply src/Files.App/Views/Settings/ActionsPage.xaml:415]: https://github.com/files-community/Files/pull/15420#issuecomment-2118429334

File: src/Files.App/Strings/en-US/Resources.resw [hunk 4]
```diff
@@ -3736,6 +3736,9 @@
   <data name="KeybindingAlreadyUsedNotification" xml:space="preserve">
     <value>This key binding is already being used, please choose a different key binding to continue.</value>
   </data>
+  <data name="KeybindingInvalidKeyNotification" xml:space="preserve">
+    <value>The key bind you pressed cannot be used, try using a different key bind.</value>
```
Comment (yaira2) [inline src/Files.App/Strings/en-US/Resources.resw:3740]:
```suggestion
    <value>This key binding cannot be used, please try again using a different key binding.</value>
```

File: src/Files.App/Views/Settings/ActionsPage.xaml (hunk 3 above)
Comment (XTorLukas) [reply src/Files.App/Views/Settings/ActionsPage.xaml:415]: If pressed down one of this or if not localised aka it's string.Empty
```csharp
// Define invalid keys that shouldn't be processed
var invalidKeys = new HashSet<VirtualKey>
{
	VirtualKey.CapitalLock,
	VirtualKey.NumberKeyLock,
	VirtualKey.Scroll,
};
```

https://github.com/files-community/Files/assets/84145589/5b6f9aba-7829-4ce7-8673-29b1045eb05b

Comment (yaira2) [reply src/Files.App/Views/Settings/ActionsPage.xaml:415]: Thanks

**Review by yaira2 — APPROVED** [review]

**Review by 0x5bfa — APPROVED** [review]
LGTM other than two things below, code wise.
I can test these on my end this weekend.

**Review by 0x5bfa — APPROVED** [review]
Code Quality is over A+.
Looks great to me.

I haven't tested on my end but if someone did I think this can be merged.
If still opened, I'll test this weekend.

**Comment by XTorLukas** [comment]
> Code Quality is over A+. Looks great to me.

Thank you, I'm glad I'm not a useless liability here even though I've only been in this community for a short time. I'll definitely keep trying to improve the quality of the code.

**Comment by XTorLukas** [comment]
I just have an idea for the future, it would be good to merge all the work with "Key binding" into one class and split it into functions, at least now it's too much for one method 
https://github.com/files-community/Files/blob/d015e048c50baf19fa95f537aa3913c49bc4beee/src/Files.App/Views/Settings/ActionsPage.xaml.cs#L54-L143

**Comment by 0x5bfa** [comment]
Definitely.
We might want to use MS.Windows.Interactivity and MS.Window.Interaction.Core or just call ViewModel.HandleEditorInput();

**Comment by XTorLukas** [comment]
Sounds good.

**Comment by yaira2** [comment]
Will this have any effect on existing users?

**Comment by XTorLukas** [comment]
> Will this have any effect on existing users?

It won't be, I just fixed the correct display of mapped keys and added an informative teaching tip about invalid key binding

**Pull request merged**